### PR TITLE
Tedlium3 transducer stateless

### PR DESCRIPTION
**Pull request by luomingshuang**
Update to tedlium3 transducer-stateless.
The results (WERs) are

|                                    |     dev    |    test    | comment                                  |
|------------------------------------|------------|------------|------------------------------------------|
|          greedy search             | 7.19       | 6.57       | --epoch 29, --avg 16, --max-duration 100 |
|      beam search (beam size 4)     | 7.12       | 6.37       | --epoch 29, --avg 16, --max-duration 100 |
| modified beam search (beam size 4) | 7.00       | 6.19       | --epoch 29, --avg 16, --max-duration 100 |

### CODE REVIEW

**Comment by csukuangfj** [comment]
The results are better than the one listed in https://github.com/k2-fsa/icefall/pull/183#discussion_r814467933

In addition to using modified transducer in the training, are there any new changes in this PR?

File: egs/tedlium3/ASR/local/display_manifest_statistics.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+See the function `remove_short_and_long_utt()` in transducer/train.py
+for usage.
+"""
+
+import numpy as np
+from lhotse import load_manifest
+
+
+def describe(cuts) -> None:
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/display_manifest_statistics.py:32]: Please don't re-invent the wheel. Just call its counterpart from lhotse.

File: egs/tedlium3/ASR/local/compile_hlg.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/compile_hlg.py:1]: Could you replace it with a symlink to the file in librispeech/ASR/local/compile_hlg.py?
And also for the following files that are seldomly changed?
- egs/tedlium3/ASR/local/compute_fbank_musan.py
- egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py
- egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py
- egs/tedlium3/ASR/local/convert_transcript_words_to_tokens.py
- egs/tedlium3/ASR/local/generate_unique_lexicon.py
- egs/tedlium3/ASR/local/prepare_lang.py
- egs/tedlium3/ASR/local/prepare_lang_bpe.py
- egs/tedlium3/ASR/local/test_prepare_lang.py
- egs/tedlium3/ASR/local/train_bpe_model.py
- egs/tedlium3/ASR/transducer_stateless/beam_search.py
- egs/tedlium3/ASR/transducer_stateless/conformer.py
- egs/tedlium3/ASR/transducer_stateless/decoder.py
- egs/tedlium3/ASR/transducer_stateless/encoder_interface.py
- egs/tedlium3/ASR/transducer_stateless/joiner.py
- egs/tedlium3/ASR/transducer_stateless/model.py
- egs/tedlium3/ASR/transducer_stateless/subsampling.py
- egs/tedlium3/ASR/transducer_stateless/test_decoder.py
- egs/tedlium3/ASR/transducer_stateless/transformer.py

If we ever need to modify it for a specific scenario, we can just copy it and then modify the copy.

Comment (luomingshuang) [reply egs/tedlium3/ASR/local/compile_hlg.py:1]: OK, for the unchanged files, I will use symlinks.

File: egs/tedlium3/ASR/RESULTS.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+```
+export CUDA_VISIBLE_DEVICES="0,1,2,3"
+
+./transducer_stateless/train.py \
+  --world-size 4 \
+  --num-epochs 30 \
+  --start-epoch 0 \
+  --exp-dir transducer_stateless/exp \
+  --max-duration 200 \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/RESULTS.md:31]:
```suggestion
  --max-duration 200
```

File: egs/tedlium3/ASR/RESULTS.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+| modified beam search (beam size 4) | 7.00       | 6.19       | --epoch 29, --avg 16, --max-duration 100 |
+
+The training command for reproducing is given below:
+
+```
+export CUDA_VISIBLE_DEVICES="0,1,2,3"
+
+./transducer_stateless/train.py \
+  --world-size 4 \
+  --num-epochs 30 \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/RESULTS.md:28]: The tensorboard log in the following contains information only up to epoch-28. Can you upload the whole log?

File: egs/tedlium3/ASR/RESULTS.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  --beam-size 4
+
+## modified beam search
+./transducer_stateless/decode.py \
+  --epoch $epoch \
+  --avg $avg \
+  --exp-dir transducer_stateless/exp \
+  --bpe-model ./data/lang_bpe_500/bpe.model \
+  --max-duration 100 \
+  --decoding-method beam_search \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/RESULTS.md:67]:
```suggestion
  --decoding-method modified_beam_search \
```

File: egs/tedlium3/ASR/RESULTS.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  --beam-size 4
+```
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/RESULTS.md:69]: Could you upload your pre-trained model and decoding logs/results to hugging face and leave a link here pointing to the repo in hugging face?

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/tedlium3/ASR/local/compute_fbank_tedlium.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+The generated fbank features are saved in data/fbank.
+"""
+
+import logging
+import os
+from pathlib import Path
+
+import torch
+from lhotse import CutSet, Fbank, FbankConfig, LilcomHdf5Writer
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/compute_fbank_tedlium.py:32]: We have switched to chunked lilcom writer. Please see https://github.com/k2-fsa/icefall/blob/1ff6196c445c2eeb491f318448e04626af2dc22f/egs/librispeech/ASR/local/compute_fbank_librispeech.py#L31

File: egs/tedlium3/ASR/RESULTS.md (hunk 4 above)
Comment (luomingshuang) [reply egs/tedlium3/ASR/RESULTS.md:28]: The following tensorboard log is the whole log. It includes 30 epochs (from 0 to 29 epoch).
image

File: egs/tedlium3/ASR/RESULTS.md (hunk 6 above)
Comment (luomingshuang) [reply egs/tedlium3/ASR/RESULTS.md:69]: Will add it.

File: egs/tedlium3/ASR/RESULTS.md (hunk 4 above)
Comment (csukuangfj) [reply egs/tedlium3/ASR/RESULTS.md:28]: Sorry, I missed it.

File: egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--texts", type=List[str], help="The input transcripts list."
+    )
+    parser.add_argument(
+        "--unk-id",
+        type=int,
+        default=2,
+        help="The number id for the token '<unk>'.",
+    )
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py:36]:
```suggestion
```

We can get the id for `<unk>` directly from the BPE model, so there is no need to pass it here.

File: egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+
+# Copyright 2021 Xiaomi Corporation (Author: Mingshuang Luo)
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py:3]:
```suggestion
# Copyright 2022 Xiaomi Corporation (Author: Mingshuang Luo)
```

File: egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        if "<unk>" in text:
+            text_segments = text.split("<unk>")
+            id_segments = sp.encode(text_segments, out_type=int)
+            for i in range(len(id_segments)):
+                if i != len(id_segments) - 1:
+                    y_ids.extend(id_segments[i] + [unk_id])
+                else:
+                    y_ids.extend(id_segments[i])
+        else:
+            y_ids = sp.encode([text], out_type=int)[0]
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py:73]:
```suggestion
            y_ids = sp.encode(text, out_type=int)
```

File: egs/tedlium3/ASR/prepare.sh [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+  echo -e "$(date '+%Y-%m-%d %H:%M:%S') (${fname}:${BASH_LINENO[0]}:${FUNCNAME[1]}) $*"
+}
+
+log "dl_dir: $dl_dir"
+
+if [ $stage -le -1 ] && [ $stop_stage -ge -1 ]; then
+  log "Stage -1: Download LM"
+  # We assume that you have installed the git-lfs, if not, you could install it
+  # using: `sudo apt-get install git-lfs && git-lfs install`
+  [ ! -e $dl_dir/lm ] && mkdir -p $dl_dir/lm
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:62]:
```suggestion
  mkdir -p $dl_dir/lm
```

File: egs/tedlium3/ASR/prepare.sh [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+  git clone https://huggingface.co/luomingshuang/tedlium3_lm $dl_dir/lm
+  cd $dl_dir/lm && git lfs pull
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:64]: Do you need to switch back to the directory of `egs/tedlium3/ASR` after `cd $dl_dir/lm`?

File: egs/tedlium3/ASR/prepare.sh [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+  # If you want to download Tedlium 4 gram language models
+  # using the follow commands:
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:67]:
```suggestion
  # If you want to download Tedlium 4-gram language models,
  # use the following commands:
```

File: egs/tedlium3/ASR/transducer_stateless/README.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+cd egs/tedlium3/ASR
+
+export CUDA_VISIBLE_DEVICES="0,1,2,3"
+
+./transducer_stateless/train.py \
+  --world-size 4 \
+  --num-epochs 30 \
+  --start-epoch 0 \
+  --exp-dir transducer_stateless/exp \
+  --max-duration 200 \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/README.md:19]:
```suggestion
  --max-duration 200
```

File: egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+
+    def train_dataloaders(self, cuts_train: CutSet) -> DataLoader:
+        logging.info("About to get Musan cuts")
+        cuts_musan = load_manifest(
+            self.args.manifest_dir / "cuts_musan.json.gz"
+        )
+
+        transforms = []
+        if self.args.enable_musan:
+            logging.info("Enable MUSAN")
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py:184]:
```suggestion


        transforms = []
        if self.args.enable_musan:
            logging.info("Enable MUSAN")
            cuts_musan = load_manifest(
               self.args.manifest_dir / "cuts_musan.json.gz"
            )
```

File: egs/tedlium3/ASR/transducer_stateless/pretrained.py [hunk 16]
```diff
@@ -0,0 +1 @@
+../../../librispeech/ASR/transducer_stateless/pretrained.py
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/pretrained.py:1]: You have made a change in the decoder to add `unk_id`, so `pretrained.py` cannot be reused, I think. Please have a test for your `pretrained.py` and make sure it works.

File: egs/tedlium3/ASR/transducer_stateless/train.py [hunk 17]
```diff
@@ -0,0 +1,753 @@
...
+Usage:
+
+export CUDA_VISIBLE_DEVICES="0,1,2,3"
+
+./transducer_stateless/train.py \
+  --world-size 4 \
+  --num-epochs 30 \
+  --start-epoch 0 \
+  --exp-dir transducer_stateless/exp \
+  --max-duration 200 \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/train.py:29]:
```suggestion
  --max-duration 200
```

File: egs/tedlium3/ASR/transducer_stateless/train.py [hunk 18]
```diff
@@ -0,0 +1,753 @@
...
+        disables autograd.
+    """
+    device = model.device
+    feature = batch["inputs"]
+    # at entry, feature is (N, T, C)
+    assert feature.ndim == 3
+    feature = feature.to(device)
+
+    supervisions = batch["supervisions"]
+    feature_lens = supervisions["num_frames"].to(device)[: feature.size(0)]
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/train.py:397]: Is it necessary to use `[: feature.size(0)]` at the end?

File: egs/tedlium3/ASR/transducer_stateless/train.py [hunk 19]
```diff
@@ -0,0 +1,753 @@
...
+        logging.info("Loading optimizer state dict")
+        optimizer.load_state_dict(checkpoints["optimizer"])
+
+    tedlium = TedLiumAsrDataModule(args)
+
+    train_cuts = tedlium.train_cuts()
+
+    def remove_short_and_long_utt(c: Cut):
+        # Keep only utterances with duration between 1 second and 20 seconds
+        return 1.0 <= c.duration <= 20.0
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/train.py:629]: From https://github.com/k2-fsa/icefall/blob/194a4e6864a95d7e7f8aff256045e55e56f4cbfd/egs/tedlium3/ASR/local/display_manifest_statistics.py#L59

99.9% percent of data has a duration less than 16.6 seconds. Can you replace 20.0 with 17 here so that you can use a larger `--max-duration` in the training.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/tedlium3/ASR/transducer_stateless/train.py (hunk 19 above)
Comment (luomingshuang) [reply egs/tedlium3/ASR/transducer_stateless/train.py:631]: In my experiments, I use 200. And I add some words to suggest others to use 17 if they want to use a big max-duration.

File: egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+    )
+
+    return parser.parse_args()
+
+
+def convert_texts_into_ids(
+    texts: List[str],
+    unk_id: int,
+    sp: spm.SentencePieceProcessor,
+) -> List[int]:
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py:45]:
```suggestion
) -> List[List[int]]:
```

File: egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    Args:
+      texts:
+        A string list of transcripts, such as ['Today is Monday', 'It's sunny'].
+      unk_id:
+        A number id for the token '<unk>'.
+    Returns:
+      Return a integer list of bpe ids.
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py:53]:
```suggestion
      Return an integer list of bpe ids.
```

Comment (csukuangfj) [inline egs/tedlium3/ASR/local/convert_transcript_words_to_bpe_ids.py:53]: It actually returns a list-of-list.

File: egs/tedlium3/ASR/prepare.sh [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+  # If you want to download Tedlium 4-gram language models,
+  # use the following commands:
+  #wget --continue http://kaldi-asr.org/models/5/4gram_small.arpa.gz -P $dl_dir/lm/ || exit 1
+  #wget --continue http://kaldi-asr.org/models/5/4gram_big.arpa.gz -P $dl_dir/lm/ || exit 1
+fi
+
+if [ $stage -le 0 ] && [ $stop_stage -ge 0 ]; then
+  log "Stage 0: Download data"
+
+  # If you have pre-downloaded it to /path/to/LibriSpeech,
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:74]: It is copied from the librispeech recipe. Please modify it for tedlium3.

File: egs/tedlium3/ASR/prepare.sh [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+  # you can create a symlink
+  #
+  # ln -sfv /path/to/tedlium3 $dl_dir/tedlium3
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:77]: Is it tedlium3 or tedlium? The following `if` is using `tedlium`, not `tedlium3`.

File: egs/tedlium3/ASR/prepare.sh [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+    lhotse download musan $dl_dir
+  fi
+fi
+
+if [ $stage -le 1 ] && [ $stop_stage -ge 1 ]; then
+  log "Stage 1: Prepare tedlium3 manifest"
+  # We assume that you have downloaded the tedlium3 corpus
+  # to $dl_dir/tedlium3
+  mkdir -p data/manifests
+  lhotse prepare tedlium $dl_dir/tedlium3 data/manifests
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:98]: Line 79 uses `$dl_dir/tedlium` but this line uses `$dl_dir/tedlium3`. Which one is correct?
Did you re-run this script from scratch?

File: egs/tedlium3/ASR/prepare.sh [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+  lang_dir=data/lang_phone
+  mkdir -p $lang_dir
+
+  if [ ! -f $lang_dir/train.text ]; then
+    ./local/prepare_transcripts.py \
+      --lang-dir $lang_dir \
+      --manifests-dir data/manifests
+
+  cat download/tedlium3/TEDLIUM.152k.dic |
+  grep -v -w "<s>" |
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:132]: Please add 2 steps here to indent `grep`?

File: egs/tedlium3/ASR/prepare.sh [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+  mkdir -p data/fbank
+  ./local/compute_fbank_musan.py
+fi
+
+if [ $stage -le 5 ] && [ $stop_stage -ge 5 ]; then
+  log "Stage 5: Prepare phone based lang"
+  lang_dir=data/lang_phone
+  mkdir -p $lang_dir
+
+  if [ ! -f $lang_dir/train.text ]; then
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:126]: This `if` has no ending `fi`.

I am wondering whether you have run this script from stage 0 to the end from scratch.

File: egs/tedlium3/ASR/prepare.sh [hunk 27]
```diff
@@ -0,0 +1,242 @@
...
+    ./local/prepare_transcripts.py \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:127]: Where is this file? Have checked in all the files required by this script?

File: egs/tedlium3/ASR/prepare.sh [hunk 28]
```diff
@@ -0,0 +1,242 @@
...
+      --vocab-size $vocab_size \
+      --transcript $lang_dir/transcript_words.txt
+
+    if [ ! -f $lang_dir/L_disambig.pt ]; then
+      ./local/prepare_lang_bpe.py --lang-dir $lang_dir
+    fi
+  done
+fi
+
+if [ $stage -le 7 ] && [ $stop_stage -ge 7 ]; then
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:178]: For the BPE based transducer training, I think you don't need to add stage 7 to stage 9.
Stage -1 and stage 5 are not required.

Please only add things required by your current setup.

File: egs/tedlium3/ASR/transducer_stateless/README.md (hunk 14 above)
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/README.md:19]: Please fix all comments (including this one) before marking the PR as ready.

File: egs/tedlium3/ASR/transducer_stateless/decode.py [hunk 29]
```diff
@@ -0,0 +1,487 @@
...
+        --decoding-method greedy_search
+
+(2) beam search
+./transducer_stateless/decode.py \
+        --epoch 29 \
+        --avg 16 \
+        --exp-dir ./transducer_stateless/exp \
+        --max-duration 100 \
+        --decoding-method beam_search \
+        --beam-size 4
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/decode.py:36]: Please also add usage for `modified beam search`.

File: egs/tedlium3/ASR/transducer_stateless/export.py [hunk 30]
```diff
@@ -0,0 +1,252 @@
...
+
+    cd /path/to/exp_dir
+    ln -s pretrained.pt epoch-9999.pt
+
+    cd /path/to/egs/tedlium3/ASR
+    ./transducer_stateless/decode.py \
+        --exp-dir ./transducer_stateless/exp \
+        --epoch 9999 \
+        --avg 1 \
+        --max-duration 1 \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/export.py:42]:
```suggestion
        --max-duration 100 \
```

File: egs/tedlium3/ASR/transducer_stateless/pretrained.py [hunk 31]
```diff
@@ -0,0 +1,343 @@
...
+Usage:
+
+(1) greedy search
+./transducer_stateless/pretrained.py \
+        --checkpoint ./transducer_stateless/exp/pretrained.pt \
+        --bpe-model ./data/lang_bpe_500/bpe.model \
+        --method greedy_search \
+        --max-sym-per-frame 1 \
+        /path/to/foo.wav \
+        /path/to/bar.wav \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/pretrained.py:28]:
```suggestion
        /path/to/bar.wav
```

File: egs/tedlium3/ASR/transducer_stateless/pretrained.py [hunk 32]
```diff
@@ -0,0 +1,343 @@
...
+
+(2) beam search
+./transducer_stateless/pretrained.py \
+        --checkpoint ./transducer_stateless/exp/pretrained.pt \
+        --bpe-model ./data/lang_bpe_500/bpe.model \
+        --method beam_search \
+        --beam-size 4 \
+        /path/to/foo.wav \
+        /path/to/bar.wav \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/pretrained.py:37]:
```suggestion
        /path/to/bar.wav
```

File: egs/tedlium3/ASR/transducer_stateless/pretrained.py [hunk 33]
```diff
@@ -0,0 +1,343 @@
...
+
+(3) modified beam search
+./transducer_stateless/pretrained.py \
+        --checkpoint ./transducer_stateless/exp/pretrained.pt \
+        --bpe-model ./data/lang_bpe_500/bpe.model \
+        --method modified_beam_search \
+        --beam-size 4 \
+        /path/to/foo.wav \
+        /path/to/bar.wav \
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/pretrained.py:46]:
```suggestion
        /path/to/bar.wav
```

File: egs/tedlium3/ASR/transducer_stateless/train.py (hunk 19 above)
Comment (csukuangfj) [reply egs/tedlium3/ASR/transducer_stateless/train.py:631]: > And I add some words to suggest others to use 17 if they want to use a big max-duration.

Where did you add such suggestions?

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/tedlium3/ASR/local/compute_fbank_tedlium.py [hunk 34]
```diff
@@ -0,0 +1,98 @@
...
+    with get_executor() as ex:  # Initialize the executor only once.
+        for partition, m in manifests.items():
+            if (output_dir / f"cuts_{partition}.json.gz").is_file():
+                logging.info(f"{partition} already exists - skipping.")
+                continue
+            logging.info(f"Processing {partition}")
+            cut_set = CutSet.from_manifests(
+                recordings=m["recordings"],
+                supervisions=m["supervisions"],
+            ).trim_to_supervisions(keep_overlapping=False)
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/compute_fbank_tedlium.py:73]: Could you apply `trim_to_supervisions()` after extracting features? That is, move it to the line after line 87.

File: egs/tedlium3/ASR/prepare.sh (hunk 25 above)
Comment (luomingshuang) [reply egs/tedlium3/ASR/prepare.sh:132]: Do you mean I should change 
```
  cat download ..... |
  grep .... |
  grep .... |
```
to 
```
  cat download ..... |  \
    grep .... |  \
    grep .... |  \
```
?

Comment (csukuangfj) [reply egs/tedlium3/ASR/prepare.sh:132]: Yes, for better visualization.

File: egs/tedlium3/ASR/local/prepare_transcripts.py [hunk 35]
```diff
@@ -0,0 +1,95 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+This script takes as input supervisions json dir "data/manifests"
+consisting of supervisions_TRAIN.json and does the following:
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/local/prepare_transcripts.py:21]:
```suggestion
consisting of supervisions_train.json and does the following:
```

File: egs/tedlium3/ASR/prepare.sh [hunk 36]
```diff
@@ -0,0 +1,177 @@
...
+
+# We assume dl_dir (download dir) contains the following
+# directories and files. If not, they will be downloaded
+# by this script automatically.
+#
+#  - $dl_dir/tedlium3
+#      You can find data, doc, legacy, LM, etc, inside it.
+#      You can download them from https://www.openslr.org/51
+#
+#  - $dl_dir/lm
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:17]: LMs are not needed in the current stateless transducer recipe. You can add them later if they are required.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

**Comment by luomingshuang** [comment]
Fix comments.

File: egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py [hunk 37]
```diff
@@ -209,10 +210,22 @@ def train_dataloaders(self, cuts_train: CutSet) -> DataLoader:
             logging.info(
                 f"Time warp factor: {self.args.spec_aug_time_warp_factor}"
             )
+            # Judge num_frame_masks for SpecAugment according to Lhotse'version
+            num_frame_masks = (
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py:214]: I just feel this kind of usage of ternary expression in python is hard to follow.

How about
```python3
if xxx:
  num_frame_masks = 2
else:
  num_frame_masks = 10
```

Comment (luomingshuang) [reply egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py:214]: I think it's ok.

Comment (luomingshuang) [reply egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py:214]: If the change in this file is ok, and then I will modify other asr_datamodule.py files.

File: egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py [hunk 38]
```diff
@@ -210,17 +210,17 @@ def train_dataloaders(self, cuts_train: CutSet) -> DataLoader:
...
-                2
-                if (
-                    inspect.signature(SpecAugment.__init__)
-                    .parameters["num_frame_masks"]
-                    .default
-                    == 1
-                )
-                else 10
-            )
+            # Design the value of num_frame_masks according to Lhotse's version.
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py:213]:
```suggestion
            # Set the value of num_frame_masks according to Lhotse's version.
```

File: egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py [hunk 39]
```diff
@@ -210,17 +210,17 @@ def train_dataloaders(self, cuts_train: CutSet) -> DataLoader:
...
+            # In different Lhotse's versions, the default of num_frame_masks is
+            # different.
+            num_frame_masks = 10
+            num_frame_masks_parameter = inspect.signature(
+                SpecAugment.__init__
+            ).parameters["num_frame_masks"]
+            if num_frame_masks_parameter.default == 1:
+                num_frame_masks = 2
+            else:
+                num_frame_masks = 10
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py:223]:
```suggestion
```

File: egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py (hunk 37 above)
Comment (csukuangfj) [reply egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py]: > If the change in this file is ok, and then I will modify other asr_datamodule.py files.

Yes, it looks good to me.

Comment (csukuangfj) [reply egs/tedlium3/ASR/transducer_stateless/asr_datamodule.py]: But I suggest you to create a new PR for it.

File: egs/tedlium3/ASR/prepare.sh [hunk 40]
```diff
@@ -0,0 +1,156 @@
...
+    lang_dir=data/lang_bpe_${vocab_size}
+    mkdir -p $lang_dir
+    # We reuse words.txt from phone based lexicon
+    # so that the two can share G.pt later.
+    cp data/lang_phone/words.txt $lang_dir
+
+    if [ ! -f $lang_dir/transcript_words.txt ]; then
+      log "Generate data for BPE training"
+      cat data/lang_phone/train.text | cut -d " " -f 2-
+      > $lang_dir/transcript_words.txt
```
Comment (csukuangfj) [inline egs/tedlium3/ASR/prepare.sh:140]: Should this line be placed at the end of the last line?

Please re-run this script from scratch to see whether it really works.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/tedlium3/ASR/prepare.sh (hunk 40 above)
Comment (luomingshuang) [reply egs/tedlium3/ASR/prepare.sh:140]: OK， I will re-run the prepare.sh from beginning to end.

**Comment by luomingshuang** [comment]
I did a small change for prepare.sh and run all stages successfully.

**Comment by csukuangfj** [comment]
> I did a small change for prepare.sh and run all stages successfully.

Did you start to rerun it before deleting all the generated data by the last run?

**Comment by luomingshuang** [comment]
Yes.


---Original---
From: &quot;Fangjun ***@***.***&gt;
Date: Sun, Mar 13, 2022 21:50 PM
To: ***@***.***&gt;;
Cc: &quot;Mingshuang ***@***.******@***.***&gt;;
Subject: Re: [k2-fsa/icefall] Tedlium3 transducer stateless (PR #233)




  
I did a small change for prepare.sh and run all stages successfully.
  
Did you start to rerun it before deleting all the generated data by the last run?
 
&mdash;
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by luomingshuang** [comment]
I did a small change for prepare.sh. After changing, I also run the decode.py to verify this change. The results are normal.

File: egs/tedlium3/ASR/prepare.sh (hunk 23 above)
Comment (HajarMazaheri) [reply egs/tedlium3/ASR/prepare.sh:56]: @csukuangfj 
Hi,
It seems that the following command downloads **TEDLIUM v2** instead of v3: 
`lhotse download tedlium $dl_dir`
How can we download Tedlium 3 ?
